### PR TITLE
register and deregister instances with lb working fine

### PR DESCRIPTION
**Pull request by drmaas**
register and deregister instances with load balancer operations and unit tests.

### CODE REVIEW

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy [hunk 1]
```diff
@@ -135,6 +147,137 @@ abstract class OpenstackClientProvider {
...
+   * Get port from load balancer description. Openstack load balancers have no native concept of internal port,
+   * so we store in the description field of the load balancer.
+   * this may be changed in a future version.
+   * @param pool
+   * @return
+   */
+  int getInternalLoadBalancerPort(LbPool pool) {
+    int minPort = 1
+    int maxPort = (1 << 16) - 1
+    String descriptionPattern = ".*internal_port=([0-9]+).*"
```
Comment (ashleykasim) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy]: Make this final?

**Comment by drmaas** [comment]
Before we merge this - I have one small thing to add.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/instance/DeregisterOpenstackInstancesAtomicOperationConverter.groovy [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import com.netflix.spinnaker.clouddriver.openstack.deploy.converters.OpenstackAtomicOperationConverterHelper
+import com.netflix.spinnaker.clouddriver.openstack.deploy.description.instance.OpenstackInstancesRegistrationDescription
+import com.netflix.spinnaker.clouddriver.openstack.deploy.ops.instance.DeregisterOpenstackInstancesAtomicOperation
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperations
+import com.netflix.spinnaker.clouddriver.security.AbstractAtomicOperationsCredentialsSupport
+import org.springframework.stereotype.Component
+
+@OpenstackOperation(AtomicOperations.DEREGISTER_INSTANCES_FROM_LOAD_BALANCER)
+@Component("deregisterInstancesDescription")
```
Comment (emjburns) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/instance/DeregisterOpenstackInstancesAtomicOperationConverter.groovy]: should be "deregisterOpenstackInstancesDescription" - i've run into some naming conflicts with the component descriptions and aws.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/instance/RegisterOpenstackInstancesAtomicOperationConverter.groovy [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import com.netflix.spinnaker.clouddriver.openstack.deploy.converters.OpenstackAtomicOperationConverterHelper
+import com.netflix.spinnaker.clouddriver.openstack.deploy.description.instance.OpenstackInstancesRegistrationDescription
+import com.netflix.spinnaker.clouddriver.openstack.deploy.ops.instance.RegisterOpenstackInstancesAtomicOperation
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperations
+import com.netflix.spinnaker.clouddriver.security.AbstractAtomicOperationsCredentialsSupport
+import org.springframework.stereotype.Component
+
+@OpenstackOperation(AtomicOperations.REGISTER_INSTANCES_WITH_LOAD_BALANCER)
+@Component("registerInstancesDescription")
```
Comment (emjburns) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/instance/RegisterOpenstackInstancesAtomicOperationConverter.groovy]: should be "registerOpenstackInstancesDescription" - I've run into some naming conflicts with the component descriptions and aws. they don't happen all the time, but I think specifying openstack is best practice.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/instance/DeregisterOpenstackInstancesAtomicOperationConverter.groovy (hunk 2 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/instance/DeregisterOpenstackInstancesAtomicOperationConverter.groovy]: I'm following the convention we've been using so far, without issue.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/instance/RegisterOpenstackInstancesAtomicOperationConverter.groovy (hunk 3 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/instance/RegisterOpenstackInstancesAtomicOperationConverter.groovy]: I'm following the convention we've been using so far, without issue.

Comment (emjburns) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/instance/RegisterOpenstackInstancesAtomicOperationConverter.groovy]: we're split on convention - https://github.com/spinnaker/clouddriver/blob/master/clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/DeployOpenstackAtomicOperationConverter.groovy#L28

I also had trouble building and running locally with the other ones too. I got a naming conflict with aws. (for terminate instance, it conflicts with https://github.com/spinnaker/clouddriver/blob/master/clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/converters/TerminateInstancesAtomicOperationConverter.groovy#L27) 

I'm not sure why this is happening only with me though - have you seen this before?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy [hunk 4]
```diff
@@ -18,24 +18,43 @@ package com.netflix.spinnaker.clouddriver.openstack.client
...
+ * TODO handleRequest should be refactored to remove the operation parameter, as client calls
+ * made here do not necessarily pertain to operations.
+ *
+ * TODO region support will need to be added to all client calls not already using regions
+ *
  * TODO tokens will need to be regenerated if they are expired.
  */
 abstract class OpenstackClientProvider {
 
+  final int minPort = 1
```
Comment (ashleykasim) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy]: By convention, constants should be named in all caps, but I see this isn't necessarily observed in other providers in this code base...

**Comment by lwander** [comment]
LGTM, mind if I merge?

**Comment by drmaas** [comment]
@lwander pls merge - thanks!

**Pull request merged**